src/build/env.py

```diff
@@ -1,221 +1,158 @@
+"""
+Creates and manages isolated build environments.
+"""
+import contextlib
 import os
 import platform
 import shutil
 import subprocess
 import sys
 import sysconfig
 import tempfile
-import types
-import typing
-from typing import Dict, Iterable, List, Optional, Sequence, Type
-
-if sys.version_info[0] == 2:
-    FileExistsError = OSError
+from typing import Generator, Iterable, Tuple
 
 try:
     import pip
 except ImportError:  # pragma: no cover
     pip = None  # pragma: no cover
 
-_HAS_SYMLINK = None  # type: Optional[bool]
-
-
-def _fs_supports_symlink():  # type: () -> bool
-    if not hasattr(os, 'symlink'):
-        return False
-
-    if sys.platform.startswith(('aix', 'darwin', 'freebsd', 'linux')):
-        return True
-    else:
-        with tempfile.NamedTemporaryFile(prefix='TmP') as tmp_file:
-            temp_dir = os.path.dirname(tmp_file.name)
-            dest = os.path.join(temp_dir, '{}-{}'.format(tmp_file.name, 'b'))
-            try:
-                os.symlink(tmp_file.name, dest)
-                return True
-            except (OSError, NotImplementedError):
-                return False
-
-
-def fs_supports_symlink():  # type: () -> bool
-    global _HAS_SYMLINK
-    if _HAS_SYMLINK is None:
-        _HAS_SYMLINK = _fs_supports_symlink()
-    return _HAS_SYMLINK
-
 
-class IsolatedEnvironment(object):
+@contextlib.contextmanager
+def isolated_env():  # type: () -> Generator[IsolatedEnvironment, None, None]
```

Comment:
We could put this to be a class method but wasn't sure if it's a good idea to annotate a class method with a ``contextmanager``. And I don't think this class makes sense without always using ``__enter__`` right after creation.

Reply:
Can we keep the class? You can move the environment creation to `__init__`, which makes more sense. This way we could create and manage a virtual environment without ever entering the context manager, and then enter the context manager when we need it. This way we can create the environment one time and enter it multiple times.

Reply:
But that allows:

```python
env = IsolatedBuildEnv.for_x()

# here basically any method call on env is a really bad idea 
# and undefined behavior, and as such IMHO the API itself should
# not allow it rather than adding defensive error raises at runtime

with env:
	pass 
```

I feel if people want to use the same environment for multiple builds they likely want:

```python

with isolated_env() as env:
     env.install()
     for a in (a, b):
        env.build()
```

That being said can you define a use case when people want to reuse isolated build environments. If you share them across packages they are no longer specific to the package 🤔 Do we really want that? Especially now that creating an environment is very low cost (<10ms, IMHO they are better off always recreating it).

Reply:
I think we should probably remove the `for_current` functionality entirely, and replace anywhere that we use it ourselves with something that spawns a new interpreter.

Activating and then de-activating a virtual environment _during an interpreter run_ seems very likely to cause weird bugs. What happens if, for example, the environment where you've installed `build` has the latest version of `toml`, and the version in the isolated environment is pinned to an older version? If you've followed a code path where `build` (or anything using the context manager) has imported `toml`, the "isolated" environment will get the outer environment's `toml`.

You'll run into other problems if you try to make the context manager also manage the `sys.module` cache, since something like this will fail:

```python

def f() -> foo.Bar:
    import foo
    return foo.Bar(1)

with env:
    x : foo.Bar = execute_something()

if x < foo.Bar(3):
   ...
```

This will probably have references to two different `foo.Bar` classes and fail!

If there's a context manager involved, it should manage the lifetime of the environment's existence (á la `tempfile.TemporaryDirectory`), not a context in which the environment is _active_.

Reply:
We do not change anything in the running interpreter session. `for_current` sets up a virtual environment based on the current interpreter. It was added because @uranusjr wanted to create an environment from an external interpreter.

```python
x : foo.Bar = execute_something()
```
runs in the current interpreter session. If you want to do something in the environment you will have to spawn a subprocess. `pep517` does this, and that is why we do not run it in a subprocess, it runs the subprocess for us.

I do not see a different in allowing a `IsolatedEnvironment` instance to run outside the context manager. Having the environment "activated" should not have impact on the current interpreter session.